arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,2917 @@
+
+static int mpfs_ifup(struct net_driver_s *dev);
+static int mpfs_ifdown(struct net_driver_s *dev);
+#ifdef CONFIG_NETDEV_CAN_BITRATE_IOCTL
+static int mpfs_ioctl(struct net_driver_s *dev, int cmd, unsigned long arg);
+#endif
+
+/****************************************************************************
+ * Private Function
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mpfs_can_retrieve_rx_frame
+ *
+ * Description:
+ *  Retrieve CAN/CANFD frame from RX Buffer
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *  cf    - Pointer to CANFD frame structure
+ *  ffw   - Previously read frame format word
+ *
+ * Returned Value:
+ *  None
+ *
+ * Assumptions:
+ *  Frame format word is already parsed in advance and provided as 'ffw' arg
+ *
+ ****************************************************************************/
+
+static void mpfs_can_retrieve_rx_frame(struct mpfs_driver_s *priv,
+                                   struct canfd_frame *cf,
+                                   uint32_t ffw)
+{
+  uint32_t idw;
+  unsigned int i;
+  unsigned int data_wc; /* data word count */
+  unsigned int data_bc; /* data byte count */
+  unsigned int dlc;
+  unsigned int len;
+
+  /* CAN ID */
+
+  idw = getreg32(priv->base + MPFS_CANFD_RX_DATA_OFFSET);
+  if (MPFS_CANFD_FRAME_FORMAT_W_IDE & ffw)
+    {
+      cf->can_id = (idw & CAN_EFF_MASK) | CAN_EFF_FLAG;
+    }
+  else
+    {
+      cf->can_id =
+        (idw >> MPFS_CANFD_IDENTIFIER_W_IDENTIFIER_BASE_SHIFT) &
+        CAN_SFF_MASK;
+    }
+
+  /* BRS, ESI, RTR Flags */
+
+  cf->flags = 0;
+  if (MPFS_CANFD_FRAME_FORMAT_W_FDF & ffw)
+    {
+      /* Enable bitrate switch by default if frame is CANFD */
+
+      cf->flags |= CANFD_BRS;
+
+      if (MPFS_CANFD_FRAME_FORMAT_W_ESI_RSV & ffw)
+        {
+          cf->flags |= CANFD_ESI;
+        }
+    }
+  else if (MPFS_CANFD_FRAME_FORMAT_W_RTR & ffw)
+    {
+      cf->can_id |= CAN_RTR_FLAG;
+    }
+
+  /* RWCNT : RX Count of Words without FRAME_FORMAT WORD. Minus the 3 words
+   * for 1 IDW, 2 timestamp words
+   */
+
+  data_wc = ((MPFS_CANFD_FRAME_FORMAT_W_RWCNT & ffw) >>
+    MPFS_CANFD_FRAME_FORMAT_W_RWCNT_SHIFT) - 3;
+  data_bc = data_wc * 4;
+
+  /* DLC */
+
+  dlc = (MPFS_CANFD_FRAME_FORMAT_W_DLC & ffw) >>
+    MPFS_CANFD_FRAME_FORMAT_W_DLC_SHIFT;
+  if (dlc <= 8)
+    {
+     len = dlc;
+    }
+  else
+    {
+      if (MPFS_CANFD_FRAME_FORMAT_W_FDF & ffw)
+        {
+          len = data_wc << 2;
+        }
+      else
+        {
+          len = 8;
+        }
+    }
+  cf->len = len;
+  if (expect_false(len > data_bc))
+    {
+      len = data_bc;
+    }
+
+  /* Timestamp - Read and throw away */
+
+  getreg32(priv->base + MPFS_CANFD_RX_DATA_OFFSET);
+  getreg32(priv->base + MPFS_CANFD_RX_DATA_OFFSET);
+
+  /* Data */
+
+  for (i = 0; i < len; i += 4)
```

Comment:
> How does this work, you read the same address (memory mapped register?) over and over. Does reading it somehow automatically forward the FIFO also?

-> yes. by raeding RX buffer, the RX buffer read pointer is automatically increased if RX buffer is not empty